File: extract_frontend_refs.js
const fs = require('fs');
const path = require('path');

const src = path.resolve('src');
const tables = new Set();
const rpcs = new Set();
const columns = new Set();
const storageBuckets = new Set();

function scanFile(filePath) {
  const text = fs.readFileSync(filePath, 'utf-8');
  
  // .from('table')
  let m;
  const fromRe = /\.from\s*\(\s*['"`]([^'"`]+)['"`]/g;
  while ((m = fromRe.exec(text)) !== null) tables.add(m[1]);
  
  // .rpc('func')
  const rpcRe = /\.rpc\s*\(\s*['"`]([^'"`]+)['"`]/g;
  while ((m = rpcRe.exec(text)) !== null) rpcs.add(m[1]);
  
  // .storage.from('bucket')
  const storageRe = /\.storage\.from\s*\(\s*['"`]([^'"`]+)['"`]/g;
  while ((m = storageRe.exec(text)) !== null) storageBuckets.add(m[1]);
  
  // .select('col1, col2')
  const selectRe = /\.select\s*\(\s*['"`]([^'"`]+)['"`]/g;
  while ((m = selectRe.exec(text)) !== null) {
    const cols = m[1].split(',').map(c => c.trim().replace(/^\*+|\*+$/g, ''));
    cols.forEach(c => { if (c && !c.startsWith('count') && c.length > 0) columns.add(c); });
  }
  
  // .insert({col: ...}) and .update({col: ...})
  const insertUpdateRe = /\.(insert|update)\s*\(\s*\{[^}]*\b([A-Za-z_][A-Za-z0-9_]*)\s*:/g;
  while ((m = insertUpdateRe.exec(text)) !== null) columns.add(m[2]);
  
  // .eq('col', ...), .order('col'), .ilike('col'), etc.
  const queryRe = /\.(eq|order|ilike|like|neq|in|contains|overlaps|gt|gte|lt|lte|or|match|not)\s*\(\s*['"`]([^'"`]+)['"`]/g;
  while ((m = queryRe.exec(text)) !== null) {
    const col = m[2];
    if (col && col.length > 0) columns.add(col);
  }
}

function walk(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory() && entry.name !== 'node_modules' && entry.name !== 'dist' && entry.name !== '.git') {
      walk(full);
    } else if (entry.isFile() && (entry.name.endsWith('.ts') || entry.name.endsWith('.tsx'))) {
      scanFile(full);
    }
  }
}

walk(src);

// Also scan api/ and supabase/functions if needed
const apiDir = path.resolve('api');
if (fs.existsSync(apiDir)) walk(apiDir);

columns.delete('Array');
columns.delete('from');
columns.delete('key');
columns.delete('ref');

console.log('=== TABLES ===');
for (const t of [...tables].sort()) console.log(t);
console.log(`\nTotal tables: ${tables.size}`);

console.log('\n=== RPC FUNCTIONS ===');
for (const r of [...rpcs].sort()) console.log(r);
console.log(`\nTotal RPCs: ${rpcs.size}`);

console.log('\n=== STORAGE BUCKETS ===');
for (const s of [...storageBuckets].sort()) console.log(s);
console.log(`\nTotal storage buckets: ${storageBuckets.size}`);

console.log('\n=== COLUMNS ===');
for (const c of [...columns].sort()) console.log(c);
console.log(`\nTotal columns: ${columns.size}`);

File: extract_frontend_refs.py
import re, os
from pathlib import Path

src = Path('src')
tables = set()
rpcs = set()
columns = set()
storage_buckets = set()

for f in src.rglob('*.tsx'):
    if 'node_modules' in str(f): continue
    text = f.read_text(errors='ignore')
    for m in re.finditer(r"\.from\s*\(\s*['\"]([^'\"]+)['\"]", text):
        tables.add(m.group(1))
    for m in re.finditer(r"\.rpc\s*\(\s*['\"]([^'\"]+)['\"]", text):
        rpcs.add(m.group(1))
    for m in re.finditer(r"\.storage\.from\s*\(\s*['\"]([^'\"]+)['\"]", text):
        storage_buckets.add(m.group(1))
    for m in re.finditer(r"\.select\s*\(\s*['\"]([^'\"]+)['\"]", text):
        cols = m.group(1).replace(',', ' ').split()
        columns.update(c.strip('*,') for c in cols if c.strip('*,') and not c.startswith('count') and not c.startswith('*') and len(c.strip('*,')) > 0)
    for m in re.finditer(r"\.insert\s*\(\s*\{[^}]*\b([A-Za-z_][A-Za-z0-9_]*)\s*:", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.update\s*\(\s*\{[^}]*\b([A-Za-z_][A-Za-z0-9_]*)\s*:", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.eq\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.order\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.ilike\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.like\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.neq\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.in\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.contains\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.overlaps\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.gt\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.gte\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.lt\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.lte\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.or\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.match\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.not\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))

for f in src.rglob('*.ts'):
    if 'node_modules' in str(f): continue
    text = f.read_text(errors='ignore')
    for m in re.finditer(r"\.from\s*\(\s*['\"]([^'\"]+)['\"]", text):
        tables.add(m.group(1))
    for m in re.finditer(r"\.rpc\s*\(\s*['\"]([^'\"]+)['\"]", text):
        rpcs.add(m.group(1))
    for m in re.finditer(r"\.storage\.from\s*\(\s*['\"]([^'\"]+)['\"]", text):
        storage_buckets.add(m.group(1))
    for m in re.finditer(r"\.select\s*\(\s*['\"]([^'\"]+)['\"]", text):
        cols = m.group(1).replace(',', ' ').split()
        columns.update(c.strip('*,') for c in cols if c.strip('*,') and not c.startswith('count') and not c.startswith('*') and len(c.strip('*,')) > 0)
    for m in re.finditer(r"\.insert\s*\(\s*\{[^}]*\b([A-Za-z_][A-Za-z0-9_]*)\s*:", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.update\s*\(\s*\{[^}]*\b([A-Za-z_][A-Za-z0-9_]*)\s*:", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.eq\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.order\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.ilike\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.like\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.neq\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.in\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.contains\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.overlaps\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.gt\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.gte\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.lt\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.lte\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.or\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.match\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))
    for m in re.finditer(r"\.not\s*\(\s*['\"]([^'\"]+)['\"]", text):
        columns.add(m.group(1))

columns.discard('Array')
columns.discard('from')
columns.discard('key')
columns.discard('ref')

print('=== TABLES ===')
for t in sorted(tables):
    print(t)
print(f'\nTotal tables: {len(tables)}')

print('\n=== RPC FUNCTIONS ===')
for r in sorted(rpcs):
    print(r)
print(f'\nTotal RPCs: {len(rpcs)}')

print('\n=== STORAGE BUCKETS ===')
for s in sorted(storage_buckets):
    print(s)
print(f'\nTotal storage buckets: {len(storage_buckets)}')

print('\n=== COLUMNS ===')
for c in sorted(columns):
    print(c)
print(f'\nTotal columns: {len(columns)}')
